services/controlplane.go

```diff
@@ -39,6 +44,68 @@ func RunControlPlane(ctx context.Context, controlHosts []*hosts.Host, localConnD
 	return nil
 }
 
+func UpgradeControlPlane(ctx context.Context, kubeClient *kubernetes.Clientset, controlHosts []*hosts.Host, localConnDialerFactory hosts.DialerFactory, prsMap map[string]v3.PrivateRegistry, cpNodePlanMap map[string]v3.RKEConfigNodePlan, updateWorkersOnly bool, alpineImage string, certMap map[string]pki.CertificatePKI, upgradeStrategy *v3.NodeUpgradeStrategy, newHosts map[string]bool) error {
+	if updateWorkersOnly {
+		return nil
+	}
+	var drainHelper drain.Helper
+
+	log.Infof(ctx, "[%s] Processing control plane components for upgrade one at a time", ControlRole)
+	if upgradeStrategy.Drain {
+		drainHelper = getDrainHelper(kubeClient, *upgradeStrategy)
+	}
+	// upgrade control plane hosts one at a time for zero downtime upgrades
+	for _, host := range controlHosts {
+		log.Infof(ctx, "Processing controlplane host %v", host.HostnameOverride)
+		nodes, err := k8s.GetNodeList(kubeClient)
+		// not returning error from here, since we still do another call to get the current node to check its status
+		var maxUnavailableHit bool
+		if err == nil {
+			for _, node := range nodes.Items {
+				// in case any previously added nodes or till now unprocessed nodes become unreachable during upgrade
+				if !k8s.IsNodeReady(node) {
+					maxUnavailableHit = true
+					break
+				}
+			}
+			if maxUnavailableHit {
+				return err
+			}
+		}
+		upgradable, err := isControlPlaneHostUpgradable(ctx, host, cpNodePlanMap[host.Address].Processes)
```

Comment:
are we supposed to check this for newHosts as well? because otherwise, I think this would make it easier to read? 
```
if newHosts[hostName] {
    deployControl()
    continue
}
// entire flow
```